### DISCUSSION TITLE
Integrate NS-3 script with MOSAIC

### DISCUSSION THREAD

**Discussion by abdullah-traboulsi** [eclipse/mosaic#53] [Q&A]
Hi,

Is it possible to integrate (convert) an NS-3 script into a MOSAIC scenario? 
I would like to make sure that my NS-3 scripts are assigned to V2X communication.
But whenever I run Tiergarten example with NS-3 federate enabled, then it would state that it is deployed but with no NS-3 scripts running.
I understand that there is a conversion tool for converting SUMO scenario, is there a similar NS-3?

What about supporting an LTE or custom made NS-3 module? Are they supported?
I found the "Ns3Ambassador.java" that creates a new object of NS3Ambassador, with only AdHoc Coomunication. Is this supposed to be the method to write and execute NS3 scripts?

I really appreciate if there is an example scenario with ns3-simulation for communication protocol between vehicles.

I have tried running Barnim scenario while enabling NS3 communication, but I received the following in the logs "Communication.log"
`Ns3Ambassador - This V2XMessage requires a destination type (CELL_GEOCAST) currently not supported by this network simulator. Skip this message. Sender=rsu_0, Receiver=NetworkAddress{address=/255.255.255.255}, V2XMessage.id=0`

Thank you

**Comment by abdullah-traboulsi** [comment]
Hi Karl,

what is the most direct way to write my own custom class, so that I don't have to only use AdHoc communication library. 
Rather, I am interested in using another modules from NS3.

I am having trouble digesting the following relationship in Mosaic, as I can't find exactly where it is you are calling for NS3 modules:

`Mosaic Ambassador <-> NS3 Federate <-> ClientServerChannel`

**Reply by abdullah-traboulsi** [reply]
I am a rookie in my understanding of NS3 connection with MOSAIC application.

Please, if you could provide any guidance or show me if any of my statements were wrong, then it is greatly appreciated.

Thank you

**Comment by rprotzmann** [comment]
The current architecture for the MOSAIC – ns-3 coupling follows this structure:
`other simulators <--Interactions--> MOSAIC (RTI) <--Interactions--> ns-3 Ambassador <--ClientServerChannel--> ns-3 Federate`

ns-3 related Interactions are e.g. `V2xMessageTransmission`, `V2xMessageReception`, … (see `runtime.json` for subscribed interactions for ns-3)

- The ns-3 Ambassador is connected to MOSAIC and translates exchanged Interactions from/to commands towards the ns-3 Federate
- The `ClientServerChannel` realizes a socket interface and depicts the boundary between Java and C++.
- The ns-3 Federate, more specifically the [`MosaicNs3Server`](https://github.com/mosaic-addons/ns3-federate/blob/main/src/mosaic-ns3-server.cc) handles the Commands (which are closely related to the Interactions) by scheduling events for the models in the communication stack or for advancing time in ns-3 itself.

The flexibility of running own scripts is somehow limited as the `MosaicNs3Server` expects that the running models could handle all the scheduled events. Due to this, the currently implemented communication stack also supports no geo routing algorithms, but only single hop broadcasting.

Of course, we would be interested in extensions in this field as well as LTE or other custom-made ns-3 modules.

**Reply by abdullah-traboulsi** [reply]
Hi Robert,

Thanks for your reply, I would be happy to collaborate on this project.
But, I am having trouble getting MOSAIC running with NS-3.31 (which includes 5G mmWave module) that I am interested in.

Are you able to assist me in this?
Here is my email address (at666@njit.edu) for better communication.

Thanks

**Comment by eqwhip** [comment]
Hello,

I am working on a similar project : integrating sidelink lte-v2x communication (using custom ns3 model) with Eclipse mosaic.

I was wondering if there is any solution or update regarding this topic.

thanks you.
